### ISSUE TITLE
Add helper to apply a function after awaiting

### ISSUE THREAD

**Issue by mristin** [maxfischer2781/asyncstdlib#33]
This is a fork from the issue #27.
We discuss how to implement a helper that applies a function on the result after awaiting an awaitable.

@maxfischer2781 wrote:

> My minimum requirement would be to support arbitrarily many arguments. This would match the old Python2 apply quite clearly. There is some overlap with partial, but that would be pushing it.

```    
async def apply(__fnc: Callable[..., R], *args: Awaitable[T], **kwargs: Awaitable[T]) -> R:
    return __fnc(*[await arg for arg in args], **{k: await arg for k, arg in kwargs.items()})
```

> That should be fine with a fast-path if there are no kwargs, and possibly even if there is only one args.

**Comment by mristin** [comment]
@maxfischer2781 could you please explain briefly the use case with a variadic arguments? I've wanted to write a pull request, but then realized that I do not understand how you would use it (in particular, why you would expect variadic arguments of the *same* type `T`).

**Comment by mristin** [comment]
@maxfischer2781 Is `apply` the correct name? What about `await_then` to make it similar to `await_each`?

(Though when I come to think about it, `apply` actually fits the context.)

**Comment by maxfischer2781** [comment]
The name ``apply`` is correct; it's lifted from Python2, which beats making up our own one.

**Comment by maxfischer2781** [comment]
As for the annotations, it is fine to just use ``Awaitable[Any]`` for the parameters for now. The proper annotation is an overload chain:
```python3
@overload
async def apply(func: Callable[[T1], R], __args1: Awaitable[T1]) -> R: ...
@overload
async def apply(func: Callable[[T1, T2], R], __args1: Awaitable[T1], __args2: Awaitable[T2]) -> R: ...
```
but that is an unnecessary showstopper for now. Several other helpers need such overloads as well.

**Comment by maxfischer2781** [comment]
Closed by #34.
